### PR TITLE
Rails 6.1.7.6 Migration

### LINKED ISSUE #2030: Upgrade Rails to 6.1.7.6

### PR DESCRIPTION
**Pull request by damianhxy**
### Summary generated by Reviewpad on 02 Jan 24 17:28 UTC
This pull request includes changes to multiple files:
1. `submissions_controller.rb`: The changes focus on the `create` action. The conditional check for the presence of the `file` parameter has been updated using the `present?` method.
2. `config.ru`: The require statement for the environment file has been updated, and the application has been configured to run using Rails.
3. `config/boot.rb`: The single quotes around `'bundler/setup'` have been replaced with double quotes.
4. `.github/workflows/rubyonrails.yml`: The version of the "ruby/setup-ruby" action has been updated, and the "bundler-cache" and "cache-version" parameters remain the same.
5. `config/application.rb`: Changes include updates to import statements, URL configuration, LTI configuration, and handling of the secret key.
6. `app/views/assessments/show.html.erb`: This file includes changes to link titles.
7. Various changes in different files related to parentheses, whitespace, configuration options, and error handling.

Let me know if you need further assistance with this diff!

## Description

Migrate Autolab to Rails 6.1.7.6 (by referencing #1552)

Procedure
1. Change `rails` to 6.1.7.6 in `Gemfile`
2. Run `bundle update`
3. Run `rails app:update`

Gems with semver major changes
* `newrelic_rpm`: 8.8.0 -> 9.6.0
* `oauth2`: 1.4.10 -> 2.0.9
* `public_suffix`: 4.0.7 -> 5.0.4
* `rack-protection`: 2.2.0 -> 3.2.0
* `rspec-rails`: 5.1.2 -> 6.1.0
* `tzinfo` 1.2.11 -> 2.0.6

### Other changes/fixes
1. Filter out more sensitive variables (supersedes https://github.com/autolab/Autolab/commit/adc326fff7e36e43b10abe3f0fc4cbc563094b4b, https://github.com/autolab/Autolab/commit/a61ae5f816931da93a429540072b3bd7389b0854, https://github.com/autolab/Autolab/commit/717012c54c1f36c7e257bb5b939dd1dee2c0cc5e)

2. Roster export
Previously: roster export would export the course's name for every user's course_number column. This is incorrect behavior.
Now: Correctly exports each user's course_number column

3. Drop warning
Previously: a student that dropped the course would see a huge warning at the top of the screen.
Now: Make use of flash to make it less glaring
Screenshot 2024-01-02 at 17 40 58

4. Submission errors
Previously: submissions that failed would do so silently
Now: errors display correctly via flash
Screenshot 2024-01-02 at 17 42 35

5. Invalid assessment warnings
Previously: all users (incl students) would see the warnings
Now: only instructors will see the warnings
Screenshot 2024-01-02 at 17 43 30

6. Gradebook search bar
Previously: CAs gradebook search bar would be flushed right
Now: search bar is correctly flushed left

7. CAs viewing section gradesheet
Previously: when filtering by section, students from other lectures would show up. Also, all students with missing submissions are displayed, not just those from the CA's lecture/section.
Now: only students from CA's section/lecture show up.

**Note:** Filtering by lecture AND section should be the expected behavior, because that's the criteria used when releasing section grades / downloading section submissions.

8. Statistics page
Previously: accessing statistics page results in error if course config can't be loaded
Screenshot 2024-01-03 at 01 31 19
Now: statistics page renders, and displays error message
Screenshot 2024-01-03 at 00 54 58

9. Misc
Remove unused "escape" param when filtering gradesheet by section
Remove unused `:viewGradesheet2` `action_auth_level`

## Motivation and Context

Resolve dependabot alerts and pave the way for a Ruby upgrade.

Closes #2030

## How Has This Been Tested?

Automated tests + [Migration tests](https://docs.google.com/spreadsheets/d/1dedvmzBFWrrRw0z9iSnj0czBbkZqZUU2ZU0F-7WYqhY/edit#gid=0)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://docs.autolabproject.com/)
- [ ] I have updated the documentation accordingly, included in this PR

### CODE REVIEW

File: app/controllers/assessment/grading.rb
Comment (damianhxy) [inline app/controllers/assessment/grading.rb:1]: This whole file is marked as a diff due to the stripping of `^M` (carriage-return) characters.

To more effectively review it, click the settings icon above and select "hide whitespace"

**Comment by 20wildmanj** [comment]
Make sure you're adding necessary `require` to each file that uses lib files, I got an error when trying to run the `autolab:populate` task because `Archive` wasn't loaded in `assessment.rb`

**Review by 20wildmanj — APPROVED** [review]
tiny nits i found here unrelated to release: https://docs.google.com/spreadsheets/d/1dedvmzBFWrrRw0z9iSnj0czBbkZqZUU2ZU0F-7WYqhY/edit#gid=1532830052

otherwise lgtm besides making sure library files are loaded properly

**Comment by damianhxy** [comment]
Fixed nits + updated error handling for statistics page (see PR description)

**Pull request merged**